### DISCUSSION TITLE
Audio quality for streams-sd-mp3-i2s

### DISCUSSION THREAD

**Discussion by petahl** [pschatzmann/arduino-audio-tools#1421] [Q&A] [answered]
I am developing an application based on the example streams-sd-mp3-i2s. I had no problems to get it work but after also testing streams-url_mp3-i2s I am a bit disappointed in the audio quality when streaming from SD card. The audio quality when streaming from url is fantastic, crystal clear. My test mp3 files sound nice when played from iTunes using the same speakers.
What can I improve to get best possible audio quality when streaming from SD card?

My equipment: Az-delivery ESP-32 Dev Kit C V4 and SPI SD card adapter (looks like the one shown here: https://www.pschatzmann.ch/home/2021/11/02/streaming-mp3-files-to-a-bluetooth-speaker/). DAC: UDA-1334 (copy). Speakers JBL Creature.

**Comment by pschatzmann** [comment, marked as answer]
I guess you are using libhelix and you just ignored the information in the Readme of that project related to the bitrates.

**Reply by petahl** [reply]
I guess you refer to the following edits I made in the MP3DecoderHelix.h:
#define MP3_MAX_OUTPUT_SIZE 2048 * 5  //1024 * 5
#define MP3_MAX_FRAME_SIZE 3200  //1600 
I also tried to maximize the i2s buffer
#define I2S_BUFFER_SIZE 1024
I also converted one mp3 file to 128kBs but did not hear so much difference.
When I create mp3 files for the SD card, what bitrate is recommended?

**Comment by pschatzmann** [comment]
I never had any issue with what Audacity classifies as Standard Speed (170-210). If you go higher, you need to adjust the buffer as descirbed in the Readme.

If this is not helping you need to double check and maybe increase the SPI speed, but usually Wifi streaming is slower then SD reading. You can try out some of the measuring sketches to compare the thruput..

**Reply by petahl** [reply]
Thank's a lot for your advice. I will do some measuring. 
Woring with audio-tools is fantastic!

**Comment by petahl** [comment]
Eventually I ended up with quite OK sound quality after the following actions:
1.MP3DecoderHelix.h:
#define MP3_MAX_OUTPUT_SIZE 2048 * 5 //1024 * 5
#define MP3_MAX_FRAME_SIZE 3200 //1600
2. I also maximized the i2s buffer using a separate AudioConfigLocal.h file
#define I2S_BUFFER_SIZE 1024
3. From the site https://forum.arduino.cc/t/sd-card-benchmark-with-esp32/987413 I found how to initialize the SD card module at different frequencies. Obviously 4MHz is default but my module is able to work at 16MHz.
4. Taking advantage of the latest update from PS of the streams-sd_mp3-i2s example.
5.  Not using the AudioLogger in the final version.

Streaming a 231 Kbit/s (VBR) mp3 file was after the above actions a much nicer experience.
